### PR TITLE
fix(perf): resume semantic preparation from durable source windows

### LINKED ISSUE #1710: Make long-session resume O(delta): persist a source transcript frontier and active model window
Parent investigation: #1708

## Problem

A real resumed OpenCode session showed:

```text
13:37:39.461  turn received: messages=5580
13:37:39.831  max_tokens resolved
13:38:53.865  ltm-budget
13:38:55.690  gradient complete: 180409 tokens
13:38:55.710  first upstream request
```

Lore therefore spent about **74 seconds after request parsing/session identification but before LTM selection**, on a known Tier-1 session. The gradient itself took only about 1.8 seconds. OpenCode reconstructing/sending the transcript is real overhead, but it cannot explain this measured Lore-side interval.

The pathological dimension appears to be structural history size (`5580` messages), not only tokens (`180409`).

## Current hot path

On every normal turn, including a resumed known session, Lore currently rebuilds the full semantic transcript:

1. `gatewayMessagesToLore(req.messages, sessionID)` over every source message.
   - SHA-256 over complete message content for the current deterministic ID.
   - A second complete-content hash for the legacy ID.
   - Another hash per part.
   - Part/object allocation for the entire transcript.
   - BPE work for provenance-bearing messages.
2. `responsesProvenanceByMessageId(...)` over the full history.
3. `resolveToolResults(...)` makes three passes over all messages/parts.
4. Tool-result-only messages invoke `temporal.storedMessageId(...)`; that performs a synchronous SQLite compatibility lookup. A tool-heavy coding transcript can therefore issue a large N+1 query set on the pre-upstream main-thread path.
5. Post-response temporal bookkeeping can rebuild/resolve the full transcript again even though it only persists the latest user message and new assistant response.

`gatewayMessagesToLore` already has a `startIndex` argument, but the pipeline currently calls it with the complete history.

## Design goal

For an append-only resumed session, semantic preparation should be:

```text
O(new/changed source suffix + bounded active raw window)
```

—not `O(all historical messages)`.

Request-body decoding may remain `O(wire bytes)` until #TBD-streaming lands, but hashing, BPE, object construction, tool pairing, DB identity resolution and gradient input construction must become suffix/window bounded.

## State model

Keep two distinct durable concepts. Do not overload gradient's `lastKnownMessageCount`, which is the **compressed model-window count**, not the source transcript frontier.

### 1. Source transcript frontier

Answers: “Which client-owned transcript prefix has Lore already validated and ingested?”

Suggested persisted fields:

```ts
interface SourceFrontier {
  sourceMessageCount: number;       // normalized source-message count
  sourceBoundaryDigest: string;     // digest of a small canonical boundary window
  sourceHeadId?: string;            // trusted client head/revision when available
  sourceProtocol: GatewayProtocol;
  adapterVersion: number;
}
```

### 2. Active model-window checkpoint

Answers: “What bounded semantic state is needed to construct the next request?”

It should preserve enough state to merge a source suffix without replaying all history:

- active raw/source messages still eligible for the model window;
- stable Lore message IDs and source indexes;
- retained Responses provenance needed for lossless encrypted-reasoning/opaque-item replay;
- unresolved tool calls crossing the frontier;
- distillation/prompt-delta generation/version;
- token estimate.

The exact representation is open: a versioned compressed `LoreMessageWithParts[]` checkpoint plus retained provenance is likely the simplest correctness-first implementation. A later iteration may reconstruct the bounded window from richer temporal/tool tables instead.

## OpenCode fast-path metadata

Lore's OpenCode `chat.headers` hook receives the current `UserMessage`, including its stable `message.id`. OpenCode also invokes `experimental.chat.messages.transform` on the full typed message list before LLM request preparation, and `chat.headers` runs later.

Use those hooks to cache and send a compact trusted revision hint, for example:

```text
x-lore-source-head-id
x-lore-source-message-count
x-lore-source-revision
```

The gateway must still have a conservative fallback for generic clients. Stable client metadata permits an O(1) append-continuation decision; absent that, validate a canonical boundary window or stream-compute a prefix digest and fall back to full reconciliation on mismatch.

### Phase A — acute hot-path fixes and proof

- Add spans/timers and counters for:
  - `gatewayMessagesToLore`
  - provenance map construction
  - `resolveToolResults`
  - `storedMessageId` call/query count
  - post-response conversion/storage
- Record `source_messages`, `source_parts`, `tool_results`, `provenance_bytes`, and `semantic_prep_ms`.
- Stop rebuilding the full transcript in post-response bookkeeping; pass the already-built representation or only convert the latest user message.
- Add a batched `storedMessageIds(...)` resolver, or complete the pre-v82 identity migration so compatibility lookup is not paid once per historical tool result.

### Phase B — durable source frontier

- Persist the dedicated source frontier.
- On a known session, validate append-only continuation.
- Convert only `req.messages.slice(frontier.sourceMessageCount)` with the correct absolute `startIndex` plus a minimal boundary overlap when required.
- Persist advancement in the same transaction as temporal writes/checkpoint update.

### Phase C — bounded active-window checkpoint

- Rehydrate the checkpoint on restart.
- Merge only the validated suffix and unresolved cross-boundary tool state.
- Run gradient from distillations + bounded active raw window, not the entire client transcript.
- Save the next checkpoint after a successful turn.

## Correctness and fallback conditions

Abandon the incremental path and perform a full reconciliation when any of these occurs:

- source count decreases or the boundary/revision does not match;
- client compaction, rewind, branch or historical edit is detected;
- protocol or canonical adapter version changes;
- checkpoint/frontier is missing, corrupt or transactionally inconsistent;
- session/project/tenant identity changes;
- retained Responses provenance cannot be reconstructed losslessly.

Full fallback is a safety mechanism, not an error.

Frontier advancement must be atomic with durable ingestion. A crash may cause harmless suffix replay, but must never leave Lore claiming a prefix was ingested when its rows/checkpoint were not committed.

## Required tests

- 5,580-message / ~180k-token representative fixture, warm and first request after restart.
- Append-only next turn processes only the suffix.
- Tool result resolves a tool call on the checkpoint boundary.
- Parallel tool calls/results preserve ordering and status.
- OpenAI Responses encrypted reasoning/opaque provenance remains byte/semantically equivalent.
- Recall markers and durable knowledge delta survive restart.
- Client compaction/count drop triggers safe full fallback.
- Historical mutation/branch mismatch triggers safe full fallback.
- Crash between upstream response and checkpoint commit causes idempotent replay, not loss.
- Existing pre-v82 temporal identities remain resolvable without N+1 queries.

## Success criteria

- Pre-upstream semantic-preparation work is bounded by suffix + active window on a valid resume.
- The 5,580-message cold-resume fixture improves by at least an order of magnitude from the observed baseline, subject to a measured benchmark committed with the change.
- Synchronous SQLite query count on the pre-upstream path is bounded by the suffix/batch count, not historical tool-result count.
- Transformed request and temporal/recall behavior remain equivalent to the full-rebuild path.
- Sentry makes incremental hit, fallback reason, source/suffix/window sizes and stage latency observable.

### PR DESCRIPTION
**Pull request by BYK**
Long resumed sessions still rebuild the complete semantic transcript after the earlier batching and provenance-token fixes. This PR persists an accepted source frontier and a bounded raw window, so unchanged resumes convert no historical messages and appended turns convert only their suffix.

- Restore stable source indexes, legacy temporal identities, lossless Responses provenance, pending tools and token/calibration metadata after restart.
- Keep full-prefix wire validation for generic clients and use the full rebuild when history, protocol, ownership, tool boundaries, calibration or window bounds are uncertain.
- Publish the checkpoint in the response temporal savepoint, with generation/revision checks and project/session/tenant lifecycle invalidation. Checkpoints are bounded, local and excluded from sync.
- Add numeric hit/fallback/window telemetry and committed benchmark samples.

On the 5,580-message synthetic fixture with a roughly 170,000-token model window, preparation plus gradient takes about **75 ms after database reopen versus 230 ms for the current full rebuild**. Both paths produce identical serialized model messages. Appending two messages converts and newly estimates only those two. This comparison excludes HTTP parsing, LTM/embedding retrieval and upstream inference; reopen is in the same process. Generic validation remains O(wire bytes).

See [benchmark report](https://github.com/BYK/loreai/blob/fix/durable-source-frontier/quality/benchmarks/source-frontier-1710.md) for samples, limits and reproduction.

Validation:
- **Independent correctness and security reviews are complete; all findings are resolved.** Both conclude MERGE subject to hosted gates.
- Typecheck, lint, formatting and gateway bundle pass. Combined focused run before the Seer follow-up: **65 passed, one opt-in benchmark skipped**. The offset-zero follow-up passes **39 affected tests, one benchmark skipped**. Separate final affected runs: correctness 324 passed, security 306 passed.
- Both reviewers ran the full suite on the initial candidate: **9,928 passed, 229 skipped, nine failures**. Seven failures were reproduced on the base/environment (installer ps/proc, compact endpoint, optional CUDA asset, cache-stability); the other two were stale schema-version assertions fixed and retested here. This is not a claim of a green final hosted full suite.
- Ten consecutive sync-property runs passed independently in each review. Mutation checks catch history/tool-boundary/revision defects, pinned/plain/nuclear exhaustion, stale cumulative and changed-call estimates, and ownership/no-store/decompression guards.
- A 40-scenario gradient parity battery and a database-reopen regression preserve canonical recall results, distillation context and durable knowledge deltas exactly. Real buffered and streamed pipeline resumes also pass.
- Final published head: `450f1c90bea07f546fc897f17acc012bb2bdeaf4`; tree `ee325ecf012c2828c0bf36fb0d4538213eb4b675` matches the clean locally verified tree. The implementation and subsequent behavior-preserving offset-zero clarification/tests were independently reviewed.
- Seer's offset-zero finding was independently verified as a false positive and resolved with four new regressions, explicit comments and an equivalent positive-offset predicate. Applying the suggested change fails encrypted-provenance tests; changing the related force guard fails durable-escalation tests.
- Hosted CI and Seer are rerunning on the final head.

Fixes #1710

### CODE REVIEW

**Review by BYK — COMMENTED** [review]
Independent review summary for final head `63b72c1544f72370d9c6b91cf8d7e3962a48ce5c` (tree `cd503f929f30f47539ba2e7b2448e18fd3ad6baa`).

**Overall: MERGE subject to hosted gates.** Separate correctness and security agents completed their reviews. All findings are resolved. The final production files exactly match reviewed implementation `980896f942367cd7b9af05ff40572c6c82398960`; subsequent changes add reviewed tests and clarify benchmark wording.

| Area | Verdict | Evidence |
| --- | --- | --- |
| Full/window equivalence | PASS | Forty mixed-tool/calibration/budget scenarios; exact output and cache-size equality on successful incremental transforms. |
| Durable replay | PASS | Real canonical recall expansion, stored recall results, distillation prefix and durable knowledge delta survive SQLite reopen with exact full-path output equality. |
| Fallback and state rollback | PASS | Pin/plain/nuclear tests establish real branch preconditions; deleting each exhaustion guard fails its regression. Forced-layer and urgent state survive speculative fallback. |
| Late tool completion | PASS | Original stale-prefix defect reproduced as1213 versus1846 tokens; overlap rebuilding and affected-call estimate reuse have separate killed mutants. |
| Ownership | PASS | Source-only project/credential rebind and mismatched new/read scopes reproduced before repair. Same-statement read/insert predicates and rebind invalidation now pass; each guard is mutation-proven. |
| Atomicity/privacy/bounds | PASS | Same-transaction claim/publication; stale generation/revision/tenant rejection; no-store/amnesia/clear/merge/delete coverage; bounded inflate and local unsynced private payloads. |
| Benchmark interpretation | PASS | About75ms versus230ms for retained full rebuild, with a170k model window. Same-process reopen, generic O(wire bytes) validation and excluded costs are explicitly documented. |

Each reviewer ran the initial full suite (9,928 passed,229 skipped,nine failures), reproduced seven failures on fresh base/environment, and verified the two schema assertions were fixed. Final affected runs passed324 tests for correctness and306 for security; the final combined focused run passed65 with one opt-in skip. Bundle/typecheck/lint/format pass. Each review verified ten consecutive sync-property runs. All eight correctness and eight security mutation experiments were reverted; reviewed worktrees were hash-verified clean.

Hosted tests, platform smoke jobs, semantic lint and Seer still need to pass on the final head.

File: packages/gateway/src/source-checkpoint.ts [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+
+  get offset(): number {
+    return this.base?.window.offset ?? 0;
+  }
+  get convertedFrom(): number {
+    return this.base?.sourceCount ?? 0;
+  }
+  get sourceWindow(): SourceWindow | undefined {
+    return this.base?.window.offset ? this.base.window : undefined;
+  }
```
[the remark this answers is not in the document]
Comment (BYK) [reply packages/gateway/src/source-checkpoint.ts:260]: Verified as a false positive and clarified in 450f1c90bea07f546fc897f17acc012bb2bdeaf4.

`sourceWindow` describes an omitted prefix for gradient; it does not control checkpoint availability. `checkpoint.base` and `convertedFrom` independently restore cached messages/provenance/IDs and convert only the suffix. At offset zero the complete source is retained, so gradient needs no omitted-prefix metadata. Checkpoint finish/claim/publication also remain enabled.

Four new regressions pass on the original behavior: warm and SQLite-reopen offset-zero requests convert/estimate zero old messages, appended requests convert only two new messages, publication advances successfully, encrypted Responses provenance survives, and forced layers 2/4 are durably consumed. The focused suite passes 39 tests (one benchmark skip); typecheck/lint/format/bundle pass. An independent reviewer confirmed the diagnosis and follow-up.

I also tested the proposed non-null getter change in an isolated worktree: both warm/reopen regressions fail because the pipeline's `!sourceWindow` complete-source gate stops restoring encrypted provenance. Changing the related force guard from an offset check to a presence check leaves forced escalation persisted; both forced-layer tests catch that. All mutations were reverted.

The getter now explicitly tests `offset > 0`, with comments explaining this contract; runtime behavior is preserved.
